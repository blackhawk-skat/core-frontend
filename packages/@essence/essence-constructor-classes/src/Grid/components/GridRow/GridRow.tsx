import * as React from "react";
import {IClassProps, IRecord, ICkId} from "@essence-community/constructor-share";
import {reaction} from "mobx";
import {IGridModel} from "../../stores/GridModel/GridModel.types";
import {BaseGridRow} from "../BaseGridRow";
import {GridDetail} from "../GridDetail";

interface IGridRowProps extends IClassProps {
    store: IGridModel;
    record: IRecord;
}

export const GridRow: React.FC<IGridRowProps> = React.memo(function GridRowMmeo(props) {
    const {store, record, ...classProps} = props;
    const [isExpanded, setIsExpanded] = React.useState(false);

    React.useEffect(() => {
        if (classProps.bc.detail) {
            const handleCheckExpanded = (): boolean => {
                return Boolean(store.recordsStore.expansionRecords.get(record[store.recordsStore.recordId] as ICkId));
            };

            if (handleCheckExpanded()) {
                setIsExpanded(true);
            }

            return reaction(handleCheckExpanded, setIsExpanded);
        }

        return undefined;
    }, [classProps.bc.detail, record, store]);

    return (
        <React.Fragment>
            <BaseGridRow store={store} record={record} {...classProps} isExpanded={isExpanded} />
            {isExpanded && classProps.bc.detail ? (
                <BaseGridRow store={store} record={record} {...classProps} isDetail isExpanded>
                    <GridDetail store={store} record={record} {...classProps} />
                </BaseGridRow>
            ) : null}
        </React.Fragment>
    );
});
