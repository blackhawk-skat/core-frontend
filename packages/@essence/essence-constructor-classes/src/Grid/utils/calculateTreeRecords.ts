import {IRecord, ICkId} from "@essence-community/constructor-share/types";
import {IGridModel} from "../stores/GridModel/GridModel.types";

export function calculateTreeRecords(store: IGridModel, record: IRecord, nesting = 1): IRecord[] {
    return store.recordsStore.expansionRecords.get(record[store.recordsStore.recordId] as ICkId) &&
        store.recordsStore.recordsTree[record[store.recordsStore.recordId] as ICkId]
        ? store.recordsStore.recordsTree[record[store.recordsStore.recordId] as ICkId].reduce(
              (acc, rec) => [...acc, ...calculateTreeRecords(store, rec, nesting + 1)],
              [{...record, nesting}],
          )
        : [{...record, nesting}];
}
