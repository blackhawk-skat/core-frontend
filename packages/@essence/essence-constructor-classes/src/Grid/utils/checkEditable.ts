import {IBuilderMode} from "@essence-community/constructor-share/types";

/**
 * Проверяем на возможность редактирования
 *
 * @param {string} mode 1 - создание, 2 - редактирование
 * @param {string} editmode all - все, insert - создание, update - редактирование
 *
 * @returns {boolean} возвращает true или false
 */
export function checkEditable(mode: IBuilderMode, editmode?: string): boolean {
    switch (true) {
        case editmode === "all":
        case mode === "1" && (editmode === "insert" || editmode === "insert-editing"):
        case mode === "2" && (editmode === "update" || editmode === "update-editing"):
            return true;
        default:
            return false;
    }
}
