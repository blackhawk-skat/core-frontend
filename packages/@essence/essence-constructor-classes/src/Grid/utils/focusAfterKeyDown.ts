import {KEY_ARROW_DOWN, KEY_ARROW_LEFT, KEY_ARROW_RIGHT, KEY_ARROW_UP, QUERY_GRID_ELEMENT} from "../constants";

function isDisabled(el: HTMLButtonElement) {
    return el && el.disabled;
}

function handleFocusRow(tr: HTMLTableRowElement, isNext: boolean) {
    const cells = tr.querySelectorAll(QUERY_GRID_ELEMENT);
    const indexCell = Array.prototype.findIndex.call(cells, (el: HTMLButtonElement) => el === document.activeElement);
    const nextTr = isNext ? tr.nextElementSibling : tr.previousSibling;

    if (nextTr instanceof HTMLTableRowElement) {
        nextTr.click();

        const nextCells = Array.from(nextTr.querySelectorAll(QUERY_GRID_ELEMENT));
        const nextCell = isDisabled(nextCells[indexCell])
            ? nextCells.find((el) => !isDisabled(el))
            : nextCells[indexCell];

        if (nextCell) {
            nextCell.focus();
        } else {
            nextTr.focus();
        }
    }
}

function handleFocusCell(tr: HTMLTableRowElement, isNext: boolean) {
    const cells = Array.from(tr.querySelectorAll(QUERY_GRID_ELEMENT));
    const indexCell = cells.findIndex((el) => el === document.activeElement);
    const nextCell = isNext ? cells[indexCell + 1] : cells[indexCell - 1];

    if (nextCell) {
        nextCell.focus();
    }
}

export function focusAfterKeyDown(tr: HTMLTableRowElement, keyCode: number) {
    switch (keyCode) {
        case KEY_ARROW_RIGHT:
            return handleFocusCell(tr, true);
        case KEY_ARROW_LEFT:
            return handleFocusCell(tr, false);
        case KEY_ARROW_DOWN:
            return handleFocusRow(tr, true);
        case KEY_ARROW_UP:
            return handleFocusRow(tr, false);
        default:
            return null;
    }
}
