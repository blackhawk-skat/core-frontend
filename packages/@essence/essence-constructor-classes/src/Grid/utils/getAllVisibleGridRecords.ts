import {IGridModel} from "../stores/GridModel/GridModel.types";
import {getTreeRecords} from "./getTreeRecords";

export function getAllVisibleGridRecords(store: IGridModel) {
    return store.bc.type === "GRID" ? store.recordsStore.records : getTreeRecords(store);
}
