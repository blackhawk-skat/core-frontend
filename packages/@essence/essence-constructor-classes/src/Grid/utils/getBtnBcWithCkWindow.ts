import {IBuilderConfig} from "@essence-community/constructor-share/types";
import {VAR_RECORD_PAGE_OBJECT_ID} from "@essence-community/constructor-share/constants";

export function getBtnBcWithCkWindow(gridbc: IBuilderConfig, btnBc: IBuilderConfig): IBuilderConfig {
    if (btnBc.ckwindow) {
        return btnBc;
    }
    const ckWindow = gridbc.childwindow?.[0]?.ckwindow;
    const ckWindowObject = gridbc.childwindow?.[0]?.[VAR_RECORD_PAGE_OBJECT_ID];

    return {
        ...btnBc,
        ckwindow: ckWindow || ckWindowObject,
    };
}
