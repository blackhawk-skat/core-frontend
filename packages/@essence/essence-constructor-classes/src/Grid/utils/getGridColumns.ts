import {IBuilderConfig} from "@essence-community/constructor-share/types";
import {VAR_RECORD_PAGE_OBJECT_ID, VAR_RECORD_PARENT_ID} from "@essence-community/constructor-share/constants";

export function getGridColumns(gridBc: IBuilderConfig): IBuilderConfig[] {
    const {columns = [], detail} = gridBc;
    const gridColumns = columns.filter((column) => column.visible);

    if (detail && gridColumns.findIndex((col: IBuilderConfig) => col.datatype === "detail") === -1) {
        return [
            {
                [VAR_RECORD_PAGE_OBJECT_ID]: "detail",
                [VAR_RECORD_PARENT_ID]: gridBc[VAR_RECORD_PAGE_OBJECT_ID],
                datatype: "detail",
                type: "COLUMN",
            },
            ...gridColumns,
        ];
    }

    return gridColumns;
}
