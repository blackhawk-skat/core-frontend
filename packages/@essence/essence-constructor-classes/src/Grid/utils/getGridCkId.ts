import {IRecord, ICkId} from "@essence-community/constructor-share/types";
import {VAR_RECORD_ID} from "@essence-community/constructor-share/constants";
import {findSetKey} from "@essence-community/constructor-share/utils";

export function getGridCkId(params: IRecord, getglobal?: string, recordId: string = VAR_RECORD_ID): ICkId {
    if (getglobal) {
        const keys = findSetKey(getglobal, recordId);
        const values: IRecord = {};

        for (const globalKey in keys) {
            if (Object.prototype.hasOwnProperty.call(keys, globalKey)) {
                const fieldName = keys[globalKey];

                values[fieldName] = params[globalKey];
            }
        }

        return values[recordId] as ICkId;
    }

    return "";
}
