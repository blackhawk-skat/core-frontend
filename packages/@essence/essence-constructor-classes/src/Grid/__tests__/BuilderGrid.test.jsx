/* eslint-disable import/no-unresolved */
/* eslint-disable import/extensions */
// @flow
import * as React from "react";
import {TableCell, TableSortLabel} from "@material-ui/core";
import {
    VAR_RECORD_PARENT_ID,
    VAR_RECORD_PAGE_OBJECT_ID,
    VAR_RECORD_OBJECT_ID,
    VAR_RECORD_DISPLAYED,
    VAR_RECORD_NAME,
    VAR_RECORD_CV_DESCRIPTION,
} from "@essence-community/constructor-share/constants";
import gridMock from "../../../mocks/grid/grid";
import {createEmptyPageStore} from "../../stores";
import {mountWithTheme} from "../../utils/test";
import BuilderFilter from "../../Filter/BuilderFilter";
import Pagination from "../../Pagination/Pagination";
import GridRow from "../Row/GridRow";
import BuilderGrid, {BaseBuilderGrid} from "../BuilderGrid";
import BaseGridTableHeader from "../BaseGridTableHeader";
import {sleep} from "../../utils/base";

type GridPropsType = {
    bc: any,
    visible?: boolean,
};

type MountGridRetunType = {
    store: any,
    pageStore: any,
    wrapper: any,
};

const mountGrid = ({bc, visible = true}: GridPropsType): MountGridRetunType => {
    const pageStore = createEmptyPageStore({styleTheme: "dark"});
    const wrapper = mountWithTheme(<BuilderGrid pageStore={pageStore} visible={visible} bc={bc} />);
    const store: any = wrapper.find(BaseBuilderGrid).prop("store");

    return {pageStore, store, wrapper};
};

// eslint-disable-next-line max-lines-per-function
describe("BuilderGrid", () => {
    const filterBc = {
        [VAR_RECORD_CV_DESCRIPTION]: "Фильтр организаций",
        [VAR_RECORD_NAME]: "Filters Panel",
        [VAR_RECORD_OBJECT_ID]: "filter",
        [VAR_RECORD_PAGE_OBJECT_ID]: "filter",
        [VAR_RECORD_PARENT_ID]: "grid",
        childs: [
            {
                [VAR_RECORD_DISPLAYED]: "Краткое наименование организации",
                [VAR_RECORD_NAME]: "cv_short",
                [VAR_RECORD_PAGE_OBJECT_ID]: "cv_short_filter_column",
                column: "cv_short",
                datatype: "text",
                type: "IFIELD",
            },
        ],
        type: "FILTERPANEL",
    };

    it("render grid", () => {
        const {wrapper} = mountGrid({bc: gridMock});

        wrapper.update();

        expect(wrapper.find(BuilderFilter).exists()).toBeFalsy();
        expect(wrapper.exists()).toBeTruthy();

        wrapper.unmount();
    });

    it("render tree grid", () => {
        const {wrapper} = mountGrid({bc: {...gridMock, rootvisible: "true", type: "TREEGRID"}});

        expect(wrapper.find(BuilderFilter).exists()).toBeFalsy();
        expect(wrapper.exists()).toBeTruthy();

        wrapper.unmount();
    });

    it("render filter", () => {
        const {wrapper} = mountGrid({bc: {...gridMock, filters: [filterBc]}});

        expect(wrapper.exists()).toBeTruthy();

        wrapper.unmount();
    });

    it("can move to next page", async () => {
        const {store, wrapper} = mountGrid({bc: {...gridMock, filters: [filterBc]}});
        const spyOnSetPageNumberAction = jest.spyOn(store.recordsStore, "setPageNumberAction");

        await store.loadRecordsAction();

        wrapper
            .find(Pagination)
            .find({iconfont: "angle-right"})
            .first()
            .simulate("click");

        expect(wrapper.exists()).toBeTruthy();
        expect(spyOnSetPageNumberAction).toHaveBeenCalledWith(1);

        wrapper.unmount();
    });

    it("Check text structure for row", async () => {
        const {store, wrapper} = mountGrid({bc: {...gridMock, filters: [filterBc]}});

        await store.loadRecordsAction();

        wrapper.update();

        wrapper.find(GridRow).forEach((row) => {
            expect(row.text()).toMatchSnapshot();
        });

        wrapper.unmount();
    });

    it("Check text structure for header", async () => {
        const {store, wrapper} = mountGrid({bc: {...gridMock, filters: [filterBc]}});

        await store.loadRecordsAction();

        wrapper.update();

        wrapper
            .find(BaseGridTableHeader)
            .find(TableCell)
            .forEach((cell) => {
                expect(cell.text()).toMatchSnapshot();
            });

        wrapper.unmount();
    });

    it("Ckeck search action", async () => {
        const {store, wrapper} = mountGrid({bc: {...gridMock, filters: [filterBc]}});
        const spyOnLoadRecordsAction = jest.spyOn(store.recordsStore, "loadRecordsAction");

        // Submit search form
        wrapper
            .find(BuilderFilter)
            .find("form")
            .prop("onSubmit")(new Event("submit"));

        await sleep(0);

        expect(spyOnLoadRecordsAction).toHaveBeenCalledWith({selectedRecordId: undefined});

        wrapper.unmount();
    });

    it("Change sort of column", () => {
        const {store, wrapper} = mountGrid({bc: {...gridMock, filters: [filterBc]}});
        const spyOnLoadRecordsAction = jest.spyOn(store.recordsStore, "loadRecordsAction");

        expect(store.recordsStore.order).toEqual({
            direction: gridMock.orderdirection,
            property: gridMock.orderproperty,
        });

        wrapper
            .find(TableSortLabel)
            .first()
            .simulate("click");

        expect(store.recordsStore.order).toEqual({direction: "DESC", property: "cv_name_displayed"});
        expect(spyOnLoadRecordsAction).toHaveBeenCalledWith();

        wrapper
            .find(TableSortLabel)
            .first()
            .simulate("click");
        expect(store.recordsStore.order).toEqual({direction: "ASC", property: "cv_name_displayed"});

        wrapper.unmount();
    });

    it("orderproperty is required for grid", () => {
        const {wrapper} = mountGrid({bc: {...gridMock, orderproperty: ""}});

        expect(wrapper.text()).toContain("Необходимо заполнить orderproperty для дальнейшей работы таблицы");

        wrapper.unmount();
    });
});
