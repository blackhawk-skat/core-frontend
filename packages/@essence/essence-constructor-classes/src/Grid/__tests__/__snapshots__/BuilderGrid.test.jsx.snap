// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuilderGrid Check text structure for header 1`] = `""`;

exports[`BuilderGrid Check text structure for header 2`] = `""`;

exports[`BuilderGrid Check text structure for header 3`] = `"Организационно-правовая форма"`;

exports[`BuilderGrid Check text structure for header 4`] = `"Полное наименование"`;

exports[`BuilderGrid Check text structure for row 1`] = `"ООО \\"Великолепный мир\\", ООО\\"Великолепный мир\\", ООО Общество с ограниченной ответственностью"`;

exports[`BuilderGrid Check text structure for row 2`] = `"БФ \\"ДИАЛ-ЭНЕРГО\\" r\\"ДИАЛ-ЭНЕРГО\\", Общество с ограниченной ответственностью"`;

exports[`BuilderGrid Check text structure for row 3`] = `"ООО \\"Зебра\\" к2\\"Зебра\\" (формирование ликвидационной комиссии), Ликвидационная комиссия ООО"`;

exports[`BuilderGrid Check text structure for row 4`] = `"ООО \\"ПОЗИТИВ ПЛЮС\\" ООО1\\"ПОЗИТИВ ПЛЮС\\" Общества с ограниченной ответственностью 1"`;

exports[`BuilderGrid Check text structure for row 5`] = `"АНО 'a3 Олег Реакт тестОлег Реакт тест"`;
